sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala

```diff
@@ -200,7 +203,11 @@ case class AQEShuffleReadExec private(
       val numCoalescedPartitionsMetric = metrics("numCoalescedPartitions")
       val x = partitionSpecs.count(isCoalescedSpec)
       numCoalescedPartitionsMetric.set(x)
-      driverAccumUpdates += numCoalescedPartitionsMetric.id -> x
+      val numEmptyPartitionsMetric = metrics("numEmptyPartitions")
```

Comment:
can we simply calculate the metric value here by looking at the underlying shuffle stage? e.g.
```
child match {
  case s: ShuffleMapStage =>
    s.mapStats.map(stats => stats.bytesByPartitionId.count(_ == 0)).getOrElse(0)
  case => 0
}
```

Reply:
Populating the new field `numEmptyPartitions` for `AQEShuffleReadExec` looks very cumbersome and needs invasive code changes.

Reply:
Good point! refactored